howso/hyperparameters.amlg

```diff
@@ -913,4 +965,393 @@
 
 		(call !Return)
 	)
+
+	;Convert confusion matrix (assoc class -> (assoc of class -> counts)) to sparse deviation matrix (SDM):
+	;with the possible formats:
+	;  (assoc of class -> (assoc of class -> deviation))
+	;  (assoc of class -> (list of: (assoc of class -> deviation), default_class_deviation))
+	;where default_class_deviation is the deviation value to be used for any classes that didn't have any counts.
+	;Can also be output as a pair in the format of: (list SDM, unspecified_class_deviation)
+	;where unspecified_class_deviation is the deviation value to be used for a class that isn't defined in the SDM
+
+	;returns null if the the passed in confusion matrix does not have enough counts to be statistically significant.
+	;
+	;parameters:
+	; confusion_matrix:  assoc of class -> assoc of class -> predicted count for each class.
+	; feature: name of nominal feature
+	; smallest_count_threshold: number of predictions a class should have for it to have an explicit deviation value.
+	;	If the count is less than this value, the deviation will be in the combined deviation for the class. default value is 10
+	#!ConfusionMatrixToSDM
+	(declare
+		(assoc
+			confusion_matrix (assoc)
+			smallest_count_threshold 10
+		)
+
+		(declare (assoc
+			total_predictions
+				(apply "+" (values
+					(map
+						(lambda (apply "+" (values (current_value))))
+						confusion_matrix
+					)
+				))
+		))
+
+		;output null if the total count is too small
+		(if (< total_predictions smallest_count_threshold)
+			(conclude (null))
+		)
+
+		(declare (assoc
+			smallest_prob (/ 1 (+ 2 total_predictions))
+			max_accuracy (- 1 (/ 1 (+ 2 total_predictions)))
+			predictions_per_class_map
+				(map
+					(lambda (apply "+" (values (current_value))) )
+					confusion_matrix
+				)
+			leftover_confusion_matrix (assoc)
+			leftover_unknown_class_deviation (null)
+		))
+
+		(declare (assoc
+			;map of class -> accuracy of random guessing
+			rand_guess_accuracy_map (map (lambda (/ (current_value) total_predictions)) predictions_per_class_map)
+			all_classes_map
+				(compute_on_contained_entities (list
+					(query_value_masses
+						feature
+						(null)
```

Comment:
I think this needs to be a bit smarter and actually use the weight feature if there is one in-use from the analyze call. I figure everything else should work "automatically" in the case of case weights because the confusion matrix should result from weighted sampling.